server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileTransactionLog.java

```diff
@@ -0,0 +1,385 @@
+      return newFiles.toArray(new StoredTabletFile[0]);
+    }
+
+    Date getInitialDate() {
+      return Date.from(Instant.ofEpochMilli(initialTs));
+    }
+
+    int getNumTransactions() {
+      return tabletLog.size();
+    }
+
+    List<DatafileTransaction> getTransactions() {
+      return tabletLog.toList();
+    }
+
+    Set<StoredTabletFile> getExpectedFiles() {
+      return new HashSet<>(Arrays.asList(finalFiles));
+    }
+
+    boolean isEmpty() {
+      return tabletLog.isEmpty();
+    }
+
+    /**
+     * Return a human readable dump of the log
+     *
+     * @param extent The tablet extent
+     * @return A dump of the log
+     */
+    String dumpLog(KeyExtent extent, boolean clear) {
+      // first lets get a consistent set of files and transactions
+      long initialTs;
+      StoredTabletFile[] initialFiles;
+      List<DatafileTransaction> transactions;
+      StoredTabletFile[] finalFiles;
+      boolean consistent;
+      do {
+        long updateCount = this.updateCount;
+        initialTs = this.initialTs;
+        initialFiles = this.initialFiles;
+        finalFiles = this.finalFiles;
+        transactions = this.tabletLog.toList();
+        consistent =
+            (updateCount == this.updateCount && updateCount == this.tabletLog.getUpdateCount());
+      } while (!consistent);
+
+      // now clear out the log if requested
+      if (clear) {
+        clear();
+      }
+
+      // now we can build our dump string
+      StringBuilder builder = new StringBuilder();
+      SimpleDateFormat format = new SimpleDateFormat(DATE_FORMAT);
+      String initialDate = format.format(Date.from(Instant.ofEpochMilli(initialTs)));
+
+      builder.append(String.format("%s: Initial files for %s : %s", initialDate, extent,
+          new TreeSet<>(Arrays.asList(initialFiles))));
+      transactions.stream().forEach(t -> builder.append('\n').append(t.toString(format)));
+      if (!transactions.isEmpty()) {
+        builder.append("\nFinal files: ").append(new TreeSet<>(Arrays.asList(finalFiles)));
+      }
+
+      return builder.toString();
+    }
+  }
+
+  /**
+   * The max size of the log as derived from the configuration
+   */
+  private static class MaxLogSize {
+    private final int maxSize;
+
+    public MaxLogSize(AccumuloConfiguration config) {
+      maxSize = config.getCount(Property.TABLE_OPERATION_LOG_MAX_SIZE);
+    }
+
+    public int getMaxSize() {
+      return maxSize;
+    }
+  }
+
+  /**
+   * A simple implementation of a ring buffer
+   */
+  public static class Ring<T> {
+    private Object[] ring;
+    private volatile int first;
+    private volatile int last;
+
+    public Ring(int size) {
+      ring = new Object[size];
+      first = 0;
+      last = -1;
+    }
+
+    @SuppressWarnings("unchecked")
+    public T add(T object) {
+      Object removed = null;
+      if (ring.length > 0) {
+        if (size() == ring.length) {
+          removed = ring[first++ % ring.length];
```

Comment:
There may be an issue with incrementing the volatile field first like this.  The increment operator is not an atomic operation. It may be better to use Atomic and incrementAndGet() to remove ambiguity in the value.

Reply:
All modification operations are synchronized within Tablet, so this is not an issue.